tools/extra/qpafilter/qpafilter.py

```diff
@@ -0,0 +1,270 @@
+            LOG.error(f'Category {category} is not supported. Skipping')
+            continue
+
+        if not verifier.verify(key_vals):
+            LOG.error(f'{category}: verification failed.')
+            sys.exit(1)
+
+        if not filt.filter(key_vals):
+            LOG.error(f'{category}: filtering failed.')
+            sys.exit(1)
+
+
+def dump_blob(args):
+    """Given the binary from a previous 'create' command, print
+       human-readable output."""
+    pass
+
+
+def parse_args():
+    """Parses command line arguments"""
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('-l', '--log-level',
+                        choices=['debug',
+                                 'info',
+                                 'warning',
+                                 'error',
+                                 'critical'],
+                        default='info', help='log level to use')
+
+    parser.add_argument('-v', '--version', action='version',
+                        version='%(prog)s {}'.format(SCRIPT_VERSION),
+                        help='display version information and exit')
+
+    subparser = parser.add_subparsers()
+
+    create = subparser.add_parser('create', help='Create blob from QPA')
+
+    create.add_argument('file', type=argparse.FileType('r'), nargs='?',
+                        help='Input QPA report to process')
+
+    create.add_argument('-t', '--min-temp', type=float,
+                        default=90.0, help='minimum temperature')
+
+    create.add_argument('-f', '--virt-fatal-temp', type=float,
+                        default=100.0,
+                        help='virtual fatal temperature threshold')
+
+    create.add_argument('-w', '--virt-warn-temp', type=float,
+                        default=90.0,
+                        help='virtual warning temperature threshold')
+
+    create.add_argument('-o', '--override-temp', action='append',
+                        default=[],
+                        help='specify a temperature override as '
+                        '<label>:<percentage>')
+
+    create.set_defaults(func=create_blob_from_qpa)
+
+    dump = subparser.add_parser('dump',
+                                help='Convert blob to human-readable format')
+
+    dump.add_argument('file', type=argparse.FileType('r'), nargs='?',
+                      help='Input blob file')
+
+    dump.set_defaults(func=dump_blob)
+
+    return parser, parser.parse_args()
+
+
+def main():
+    """The main entry point"""
+    parser, args = parse_args()
+
+    if args.file is None:
+        print('Error: file is a required argument.\n')
+        parser.print_help(sys.stderr)
+        sys.exit(1)
+
+    LOG.setLevel(logging.NOTSET)
+    log_fmt = '[%(levelname)-8s] %(message)s'
+    log_hndlr = logging.StreamHandler(sys.stdout)
+    log_hndlr.setFormatter(logging.Formatter(log_fmt))
+    log_hndlr.setLevel(logging.getLevelName(args.log_level.upper()))
+    LOG.addHandler(log_hndlr)
+
+    args.func(args)
```

Comment:
If no subparser is called, args won't have a func attribute to call.

Reply:
It actually won't get that far, because it will fail on the args.file check above.  I do need to fix that check, though.  If the file isn't given, then args.file won't exist.